crates/oxc_codegen/src/code_buffer.rs

```diff
@@ -0,0 +1,421 @@
+        let hint = iter.size_hint();
+        self.buf.reserve(hint.1.unwrap_or(hint.0));
+        for c in iter {
+            self.print_ascii_char(c);
+        }
+    }
+
+    /// Print a sequence of bytes without checking that this buffer still
+    /// represents a valid UTF-8 string.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that, after being called, this buffer represents
+    /// a valid UTF-8 string. In practice, this means only two cases are valid:
+    ///
+    /// 1. Both the buffer and the byte sequence are valid UTF-8,
+    /// 2. The buffer became invalid after a call to [`print_byte`] and `bytes`
+    ///    completes any incomplete code points, returning the buffer to a valid
+    ///    state.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use oxc_codegen::CodeBuffer;
+    /// let mut code = CodeBuffer::new();
+    ///
+    /// // Indent to a dynamic level. Sound because all elements in this
+    /// // iterator are valid 1-byte UTF-8 code points (ASCII).
+    /// unsafe {
+    ///     code.print_unchecked(std::iter::repeat(b' ').take(4));
+    /// }
+    /// ```
+    ///
+    /// [`print_byte`]: CodeBuffer::print_byte
+    #[inline]
+    pub(crate) unsafe fn print_unchecked<I>(&mut self, bytes: I)
+    where
+        I: IntoIterator<Item = u8>,
+    {
+        self.buf.extend(bytes);
+    }
+
+    /// Convert a `CodeBuffer` into a string of source code, leaving its
+    /// internal buffer empty and finalizing the codegen process.
+    ///
+    /// It is safe to re-use a buffer after calling this method. Its contents
+    /// will be emptied out, but all memory resources are retained and in a
+    /// valid state. You may use [`String::from`] if you don't intend on
+    /// re-using the buffer. It simply calls this method and drops the
+    /// `CodeBuffer` afterwards.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use oxc_codegen::CodeBuffer;
+    /// let mut code = CodeBuffer::new();
+    /// code.print_str("console.log('foo');");
+    ///
+    /// let source = code.take_source_text();
+    /// assert_eq!(source, "console.log('foo');");
+    /// assert!(code.is_empty());
+    /// ```
+    #[must_use]
+    pub fn take_source_text(&mut self) -> String {
+        use std::mem::take;
+
+        #[cfg(debug_assertions)]
+        {
+            String::from_utf8(take(&mut self.buf)).unwrap()
+        }
+        #[cfg(not(debug_assertions))]
+        {
+            // SAFETY: `buf` is valid UTF-8 because of invariants upheld by
+            // CodeBuffer. If, for some reason, it is not, this is caused by
+            // improper use of `unsafe` printing methods.
+            unsafe { String::from_utf8_unchecked(take(&mut self.buf)) }
+        }
+    }
+}
+
+impl AsRef<[u8]> for CodeBuffer {
+    fn as_ref(&self) -> &[u8] {
+        &self.buf
+    }
+}
+impl From<CodeBuffer> for String {
+    fn from(mut printer: CodeBuffer) -> Self {
+        printer.take_source_text()
+    }
+}
```

Comment:
I'm not sure if this should delegate to `take_source_text`. Unlike in `take_source_text`, here you're consuming the `CodeBuffer`, so no need to `take` its contents.

If compiler inlines `take_source_text`, then it will *probably* see that it can elide the `take` operation. But no guarantee that it will inline, so I think better to play it safe and duplicate the code. And in release mode, this function is a no-op (just a type system cast) so all the code will disappear anyway.

This method should be marked `#[inline]` because it's a no-op.

```suggestion
impl From<CodeBuffer> for String {
    #[inline]
    fn from(printer: CodeBuffer) -> Self {
        #[cfg(debug_assertions)]
        {
            String::from_utf8(printer.buf).unwrap()
        }
        #[cfg(not(debug_assertions))]
        {
            // SAFETY: `buf` is valid UTF-8 because of invariants upheld by `CodeBuffer`
            unsafe { String::from_utf8_unchecked(printer.buf) }
        }
    }
}
```